hoomd/hpmc/ShapeUtils.h

```diff
@@ -0,0 +1,364 @@
+// Copyright (c) 2009-2022 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#ifndef _SHAPE_UTILS_H
+#define _SHAPE_UTILS_H
+
+#include <fstream>
+#include <iostream>
+#include <queue>
+#include <set>
+
+#include "ShapeConvexPolygon.h"
+#include "ShapeConvexPolyhedron.h"
+#include "ShapeEllipsoid.h"
+#include "ShapePolyhedron.h"
+#include "ShapeSimplePolygon.h"
+#include "ShapeSphere.h"
+#include "ShapeSpheropolygon.h"
+#include "ShapeSpheropolyhedron.h"
+#include "ShapeSphinx.h"
+#include "hoomd/extern/quickhull/QuickHull.hpp"
+
+namespace hoomd
+    {
+
+namespace hpmc
+    {
+
+namespace detail
+    {
+
+typedef EllipsoidParams ellipsoid_param;
+typedef PolyhedronVertices convex_polyhedron_param;
+
+template<class Shape> class MassPropertiesBase
+    {
+    public:
+    MassPropertiesBase() : m_volume(0.0), m_surface_area(0.0), m_center_of_mass(0.0, 0.0, 0.0)
+        {
+        for (unsigned int i = 0; i < 6; i++)
+            m_inertia[i] = 0.0;
+        }
+
+    virtual ~MassPropertiesBase() {};
+
+    Scalar getVolume()
+        {
+        return m_volume;
+        }
+
+    const vec3<Scalar>& getCenterOfMass()
+        {
+        return m_center_of_mass;
+        }
+
+    Scalar getDetInertiaTensor()
+        {
+        vec3<Scalar> a(m_inertia[0], m_inertia[3], m_inertia[5]),
+            b(m_inertia[3], m_inertia[1], m_inertia[4]),
+            c(m_inertia[5], m_inertia[4], m_inertia[2]);
+        Scalar det = dot(a, cross(b, c));
+        // determinant can be negative depending on order of vertices
+        if (det < 0.0)
+            {
+            det *= -1.0;
+            }
+        return det;
+        }
+
+    virtual void updateParam(const typename Shape::param_type& param, bool force = true) { }
+
+    virtual void compute()
+        {
+        throw std::runtime_error("MassProperties::compute() is not implemented for this shape.");
+        }
+
+    protected:
+    Scalar m_volume;
+    Scalar m_surface_area;
+    vec3<Scalar> m_center_of_mass;
+    Scalar m_inertia[6]; // xx, yy, zz, xy, yz, xz
+    };                   // end class MassPropertiesBase
+
+template<class Shape> class MassProperties : public MassPropertiesBase<Shape>
+    {
+    public:
+    MassProperties() : MassPropertiesBase<Shape>() { }
+
+    MassProperties(const typename Shape::param_type& shape) : MassPropertiesBase<Shape>()
+        {
+        this->compute();
+        }
+    };
```

Comment:
Why this two layer approach? This class does not seem to add anything?

Reply:
Without it, the `ShapeConvexPolyhedron` and `ShapeEllipsoid` specializations below result in compilation errors. We could just rename `MassPropertiesBase` to `MassProperties` and remove this completely, I think.

Reply:
This doesn't work, unless we are willing to replicate all the code from the base class in each specialization.